### DISCUSSION TITLE
admin token from .env written to config.json

### DISCUSSION THREAD

**Discussion by navels** [dani-garcia/vaultwarden#4758] [Q&A]
Unsure if this is a bug or there is some misunderstanding of where/how to properly set the admin token.

Installed vaultwarden via the proxmox script at https://github.com/tteck/Proxmox/.  I've seen that

1. Setting the `ADMIN_TOKEN` in `/opt/vaultwarden/.env` allows one to log into the admin panel.
2. But then if you click Save in the admin panel it writes the token to `/opt/vaultwarden/data/config.json
3. And then if you change the token in `.env` it has no effect because of the older token in `config.json`.

There is this:

https://github.com/dani-garcia/vaultwarden/blob/bb2412d0339e1da5dee99fc566a2b2aab5d2808c/.env.template#L5-L6

which clearly states that `config.json` overrides `.env` so this seems to be by design, but is awkward (edit: specifically for the admin token) and seems to render having the token in `.env` pretty useless.

**Comment by BlackDex** [comment]
That all depends. If you only configure the settings via environment variables then it is all fine.
If  you configure it via the admin interface then that is leading since you are making changes, and then which should we select?
That is why it is that when settings are saved via the admin interface overrule all other items.

You can just configure everything via your `.env` and only use the admin interface for user and org management and basic troubleshooting.

**Comment by navels** [comment]
That makes sense for everything *except* the admin token.  If you set the admin token once and then save in the admin panel and then later try to change the admin token in `.env`, it won't actually do anything and then you *have* to hand edit `config.json` in order to get back into the admin panel.

**Reply by BlackDex** [reply]
That is the same for every other config option, not only the admin_token, like `SMTP_USERNAME` or `SMTP_PASSWORD` for example.

The only thing we could do is make the admin token readonly, which would prevent it from being saved in the config file and also not editable in the admin interface.

**Reply by navels** [reply]
Yep, I understand that's how all the options work, it just makes for a very confusing flow when you change the token in `.env`.

**Reply by BlackDex** [reply]
It is very easy. Either do not use the `/admin` interface to configure your options and only use the `.env` way.

Or, have the initial very basic read-only options config in your `.env` and after that modify all non-read-only config options via the `/admin` interface, including your `ADMIN_TOKEN`.

**Reply by navels** [reply]
Easy, sure, but non-intuitive and confusing.

**Reply by BlackDex** [reply]
That might be, but it's also not something we can fix without breaking a lot of current installations.
If and when we would change this, it will be in a release where we probably break some more default behaviors.
